### ISSUE TITLE
How to debug packet drops in vpp

### ISSUE THREAD

**Issue by ljkiraly** [networkservicemesh/deployments-k8s#10154]
## Question

During an sctp traffic test running during the night packet re-transmission can be experienced.  Based on tcpdump traces some packets are dropped by vpp application in forwarder-vpp. The re-transmission can be seen with low traffic rate also (10TPS, originally tested by 100TPS traffic rate). The issue can be seen for a short time interval (1 second), there are no configuration ongoing during the test and the memory and CPU usage on the node doesn't show any increase.

What information should be collected? Which vpp commands should be useful for troubleshooting?

## Context

The topology is similar to the use-cases Kernel2Kernel and Kernel2Ethernet2Kernel. The vpp configuration is the default used by NSM (vpp-helper).

**Comment by glazychev-art** [comment]
@ljkiraly 
Could you please tell more about how to reproduce?
What utilities did you use? Is it possible to use `iperf3 --sctp`?
Can you please send an example of `tcpdump` logs where you see the problem?

**Comment by glazychev-art** [comment]
Thank you very much for the details!
Question, you mentioned Kernel2Kernel in the description - do you know if this problem is also reproduced there?

**Comment by ljkiraly** [comment]
> you mentioned Kernel2Kernel in the description - do you know if this problem is also reproduced there?

That problem was seen during stability test also, but on other cluster. I will open a separate ticket for that, but I need to collect the details relevant for NSM/vpp.

**Comment by ljkiraly** [comment]
An important detail, that the fault was seen in NSM v1.9.0. We asked some test with a forwarder-vpp 1.11 and newest vpp version they can see one re-transmission only. Note that was a debug build of forwarder without Calico patches and disabled TAP interfaces. The usage of veth instead of tap might have impact here. We asked a reproduction with latest NSM release.

**Comment by ljkiraly** [comment]
Seems that the issue can be seen in NSM 1.11.1-rc1 patched with AF_PACKET fix. Now only two connection failure has occurred, but for some reason no useful debug information or trace can be collected. We tried to reproduce the fault by iperf3 tcp connections. There were several re-transmissions in a short time interval:

```
> kubectl exec -n ns-kernel2ethernet2kernel $NSC -c iperf-server -- iperf3 -i0 -t 15 -c 172.16.1.0
Connecting to host 172.16.1.0, port 5201
[  5] local 172.16.1.1 port 55806 connected to 172.16.1.0 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-15.00  sec   479 MBytes   268 Mbits/sec  338    828 KBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-15.00  sec   479 MBytes   268 Mbits/sec  338             sender
[  5]   0.00-15.12  sec   476 MBytes   264 Mbits/sec                  receiver

```
But in this case there were packet dropped by kernel on NSM tap interface inside endpoint container (101 tx packet drop):

```
> kc exec -n ns-kernel2ethernet2kernel $NSC -c iperf-server -- ip -s l show dev nsm-1
3: nsm-1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1446 qdisc mq state UNKNOWN mode DEFAULT group default qlen 1000
    link/ether 02:fe:a3:c6:c3:62 brd ff:ff:ff:ff:ff:ff
    alias server-049cf7a4-c704-460a-9626-54efcc4454e6
    RX: bytes  packets  errors  dropped missed  mcast   
    6557293    99354    0       0       0       0       
    TX: bytes  packets  errors  dropped carrier collsns 
    524178015  359061   0       101     0       0 
```
Maybe something related to the tx queue length? Found some explanation of this behavior there: https://access.redhat.com/documentation/en-us/red_hat_openstack_platform/13/html/ovs-dpdk_end_to_end_troubleshooting_guide/high_packet_loss_in_the_tx_queue_of_the_instance_s_tap_interface#doc-wrapper

But I don't think it's the same issue as with sctp seen on customer test cluster since in that case there were no dropped packet in kernel statistics.

**Comment by glazychev-art** [comment]
@ljkiraly 
As I see, NSM 1.11.1-rc1 didn't include the AF_PACKET fix. Did you build it yourself?

**Comment by ljkiraly** [comment]
@glazychev-art Yes, I built it with the patch. Meantime testing with rc2 is ongoing. 
I built a version where TAP were disabled. Re-transmissions can be seen with veth also.

**Comment by glazychev-art** [comment]
The packet drops that you see after iperf session, you are right, it looks like it is related to the tx queue length.
So, if you set `ip link set dev <nsm-int-name> qlen 2000` on the NSC and NSE (you should deploy it with `privileged: true`), you won't see any drops.
But I don't think that this is the original problem and we should change the default value. Because it is the responsibility of the TCP layer to guarantee the delivery of packets. In addition, I didn't see any increase in performance when increasing qlen...
